### DISCUSSION TITLE
mir shared object size

### DISCUSSION THREAD

**Discussion by fabubaker** [vnmakarov/mir#304] [Q&A] [answered]
Out of curiosity, I wanted to check what the shared object size is for MIR (excluding `c2mir`). 

When building using `make`, the size is 1.5 MB.
When building using `make debug`, the size is 915 KB.

I have a couple of questions:
1. Any idea why the shared object is smaller when built with `make debug`?

2. Is it possible to get a smaller shared object size? I tried compiling with `-Os`, but that did not help.
Under "MIR project competitors" in README.md, when comparing against LibJIT, it says that MIR has an object size of 170 KB.
Is it possible to bring the size down that much?

**Comment by vnmakarov** [comment, marked as answer]
Thank you for interesting in MIR-project.

> Out of curiosity, I wanted to check what the shared object size is for MIR (excluding `c2mir`).
> 
> When building using `make`, the size is 1.5 MB. When building using `make debug`, the size is 915 KB.
> 
> I have a couple of questions:
> 
>     1. Any idea why the shared object is smaller when built with `make debug`?
> 

The both versions contain debug information. Non-debug version uses `-O3`.  Very aggressive optimizations (including inlining) are used in this case and this results in bigger code size and bigger debug information.

>     2. Is it possible to get a smaller shared object size? I tried compiling with `-Os`, but that did not help.
>        Under "MIR project competitors" in README.md, when comparing against LibJIT, it says that MIR has an object size of 170 KB.
>        Is it possible to bring the size down that much?

Yes, I guess,  it is possible.  First of all, -g can be removed and -Os should be used.  It gives 230K code (text+data) w/o c2mir.  The smaller code size can be reached by switching of unused code, e.g. if you use only API to create MIR you don't need read binary MIR and you don't need to read textual MIR.  If you don't need to save MIR, you can switch writing MIR binary to a file.  I am sure you can reach 170KB in this case.

A lot of code is used to print debug info. This code can be also switched off for a release code.  Also different macros help to switch off many optimizations in MIR-generator.

So MIR-code size can be varied considerably. Although using such macros are not documented and not tested regularly.

**Reply by fabubaker** [reply]
Oh I see, I didn't notice that `-g` was switched on for both `make` and `make debug`.

Removing `-g` and compiling with `-Os` gives me a shared object size of 317 KB, which is much better than LibJIT.

> A lot of code is used to print debug info. This code can be also switched off for a release code. Also different macros help to switch off many optimizations in MIR-generator.

> So MIR-code size can be varied considerably. Although using such macros are not documented and not tested regularly.

I'll try looking into some of these macros and document them here if I can. Thanks @vnmakarov!
